optaplanner-examples/src/main/java/org/optaplanner/examples/vehiclerouting/score/VehicleRoutingConstraintProvider.java

```diff
@@ -27,6 +27,7 @@ public Constraint[] defineConstraints(ConstraintFactory factory) {
 
     protected Constraint vehicleCapacity(ConstraintFactory factory) {
         return factory.forEach(Customer.class)
+                .filter(customer -> customer.getVehicle() != null)
```

Comment:
This makes me think.

CS handles nulls by falling forEach() instead of forEachIncludingNullVars(). However, this does not apply here, as this is a inverse relation shadow variable, and those are not included. That in turn forces these null checks all over the constraint provider.

We should think if there could be a way to avoid this. The customer in this case is "unassigned", and the constraint provider should have means of saying that, I think.

Anyway... unrelated to the PR.

Reply:
Yes, I totally agree. I am also concerned about these boiler-plate null checks and would love to see `forEach()` behave equally for both genuine and shadow entities.

We (you and I) have discussed this previously IIRC and I'm afraid we ran into some design issue. Something like it's not possible to decide in a generic way whether a shadow entity is initialized or not. Not sure.